crypto/bio/bss_dgram.c

```diff
@@ -1274,92 +1288,80 @@ static ossl_ssize_t dgram_sendmmsg(BIO *b, BIO_MSG *msg, size_t stride,
     if (num_msg == 0)
         return 0;
 
+    if (num_msg > OSSL_SSIZE_MAX)
+        num_msg = OSSL_SSIZE_MAX;
+
 #if M_METHOD != M_METHOD_NONE
     sysflags = translate_flags(flags);
 #endif
 
 #if M_METHOD == M_METHOD_RECVMMSG
     /*
-     * In the sendmmsg/recvmmsg case, the messages passed to us are chunked into
-     * one or more sendmmsg calls. We have to do this because we need to
-     * allocate our translated struct msghdr and struct iovec on the stack to
-     * support multithreaded use.
+     * In the sendmmsg/recvmmsg case, we need to allocate our translated struct
+     * msghdr and struct iovec on the stack to support multithreaded use. Thus
+     * we place a fixed limit on the number of messages per call, in the
+     * expectation that we will be called again if there were more messages to
+     * be sent.
      */
-    for (i = 0; i < num_msg; i += BIO_MAX_MSGS_PER_CALL) {
-        /* Translate all messages in this batch */
-        jlim = BIO_MAX_MSGS_PER_CALL;
-        if (jlim > num_msg - i)
-            jlim = num_msg - i;
-        for (j = 0; j < jlim; ++j) {
-            translate_msg(&mh[j].msg_hdr, &iov[j],
-                          control[j], &BIO_MSG_N(msg, i + j));
-
-            /* If local address was requested, it must have been enabled */
-            if (BIO_MSG_N(msg, i + j).local != NULL) {
-                if (!have_local_enabled)
-                    /*
-                     * If we have done at least one message, we must return the
-                     * count; if we haven't done any, we can give an error code
-                     */
-                    return num_done > 0 ? num_done : -1;
-
-                if (pack_local(b, &mh[j].msg_hdr, BIO_MSG_N(msg, i + j).local) < 1)
-                    return num_done > 0 ? num_done : -1;
-            }
-        }
+    if (num_msg > BIO_MAX_MSGS_PER_CALL)
```

Comment:
Should the BIO_MAX_MSGS_PER_CALL be a public define so the callers won't needlessly attempt to batch more messages? And document that BIO_MAX_MSGS_PER_CALL is a recommended limit on the BIO_sendmmsg call.

Reply:
I don't think we should make any specific promises about the behaviour of this API in terms of the number of messages it handles at this time. There can be other implementations like BIO_dgram_pair and third party implementations which can do what they want.

Maybe a BIO_ctrl to query a suggested batching factor would  be something worth considering, rather than a constant.

Reply:
> Maybe a BIO_ctrl to query a suggested batching factor would be something worth considering, rather than a constant.

Yep, that seems to me as something very useful for callers.